### PR TITLE
Refactor chunk and constraint handling for consistent catalog access

### PR DESCRIPTION
**Pull request by erimatnor**
This PR is part of the effort to make metadata catalog updates use a consistent API. The main change here is how chunks and chunk constraints are handled in this regard, making updates to related catalog tables use our internal catalog API and also unifying the way we deal with dimensional and inherited constraints (which previously were handled in different ways). 

As a consequence of this refactoring, permissions handling is now a lot more fine-grained, avoiding running large chunks of code under super-user permission. It also fixes permissions-related issues for creating chunks in associated schemas, which previously executed with super-user permissions. This resulted in user being able to create new schemas and tables in existing schemas without having the proper permissions.

The PR consists of three commits, and more information can be found in the individual commit messages.

One of the commits adds support for reloptions (storage parameters) on chunks, and fixes https://github.com/timescale/timescaledb/issues/353

### CODE REVIEW

File: src/hypertable.c [hunk 1]
```diff
@@ -308,3 +362,64 @@ hypertable_validate_triggers(PG_FUNCTION_ARGS)
...
+
+	if (PG_NARGS() != 2)
+		elog(ERROR, "Invalid number of arguments");
+
+	/*
+	 * If the schema name is NULL, it implies the internal catalog schema and
+	 * anyone can create chunks there
+	 */
+	if (PG_ARGISNULL(0))
+		PG_RETURN_NULL();
```
Comment (RobAtticus) [inline src/hypertable.c]: May be my PG internals ignorance, but why is this PG_RETURN_NULL instead of PG_RETURN_VOID like the one further down where it checks if it's the internal schema?

Comment (erimatnor) [reply src/hypertable.c]: No ignorance involved. It should return void, as further down.

File: src/process_utility.c [hunk 2]
```diff
@@ -721,6 +696,15 @@ process_altertable_add_constraint(Hypertable *ht, const char *constraint_name)
 	foreach_chunk(ht, process_add_constraint_chunk, &hypertable_constraint_oid);
 }
 
+static void
+process_drop_constraint_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
```
Comment (cevian) [inline src/process_utility.c]: `process_drop_chunk_constraint` better name?

File: test/expected/create_hypertable.out [hunk 3]
```diff
@@ -220,10 +240,10 @@ SELECT * FROM _timescaledb_internal.get_create_command('test_1dim');
 (1 row)
 
 \dt "test_schema".*
-                         List of relations
-   Schema    |          Name          | Type  |        Owner        
--------------+------------------------+-------+---------------------
- test_schema | test_1dim              | table | default_perm_user_2
```
Comment (cevian) [inline test/expected/create_hypertable.out]: Why this change? Do we want to switch back to `default_perm_user_2` on line 227(new line #s)

**Comment by cevian** [comment]
More general question (not necessarily for this PR): We often use C array's with capacity and num_elements as a container. Why are we not using Lists instead. Alternatively should we write a reusable ADT for these?

File: src/process_utility.c (hunk 2 above)
Comment (erimatnor) [reply src/process_utility.c]: I think the unofficial naming strategy so far for processUtility stuff is process_<cmd>_<level>, where we basically append a finer grained suffix as we process the command and descend down the table hierarchy. So, we first do a high-level `process_drop_constraint()` to check for regular vs hypertable, then when we recurse to chunks and call `process_drop_constraint_chunk()`. I think this is mostly the scheme we are using for other commands too.

File: src/process_utility.c [hunk 4]
```diff
@@ -1191,13 +1177,15 @@ process_alter_column_type_end(Hypertable *ht, AlterTableCmd *cmd)
...
 	ColumnDef  *coldef = (ColumnDef *) cmd->def;
 	Oid			new_type = TypenameGetTypid(typename_get_unqual_name(coldef->typeName));
-	CatalogSecurityContext sec_ctx;
+	Dimension  *dim = hyperspace_get_dimension_by_name(ht->space, DIMENSION_TYPE_ANY, cmd->name);
 
-	catalog_become_owner(catalog_get(), &sec_ctx);
+	if (NULL == dim)
+		return;
+
+	dimension_update_type(dim, new_type);
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/process_utility.c]: Though about it, but I'd rather keep mutations on the dimension object/table separate from any logic that follows as a result of that mutation, if that makes sense? Otherwise, I think we risk overloading the dimension type with lots of logic related to other objects.

File: src/chunk.c [hunk 5]
```diff
@@ -690,90 +762,128 @@ chunk_find(Hyperspace *hs, Point *p)
...
 	chunk_scan_ctx_destroy(&ctx);
 
 	if (NULL != chunk)
+	{
 		/* Fill in the rest of the chunk's data from the chunk table */
 		chunk_fill_stub(chunk, false);
 
+		/* Rescan the constraints to also get the inherited constraints */
+
+		/*
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c]: Hmm, this shouldn't be this way obviously. Will fix.

File: src/chunk.c [hunk 6]
```diff
@@ -294,18 +328,74 @@ chunk_collision_resolve(Hyperspace *hs, Hypercube *cube, Point *p)
...
 {
-	Oid			schema_oid = get_namespace_oid(schema, false);
 	Hyperspace *hs = ht->space;
 	Catalog    *catalog = catalog_get();
 	CatalogSecurityContext sec_ctx;
 	Hypercube  *cube;
 	Chunk	   *chunk;
-	int			i;
-
-	catalog_become_owner(catalog, &sec_ctx);
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c]: Do you know how to get the `LOCKMODE`?

File: src/chunk.c [hunk 7]
```diff
@@ -314,42 +404,45 @@ chunk_create_after_lock(Hypertable *ht, Point *p, const char *schema, const char
...
+	chunk->table_id = chunk_create_table(chunk, ht);
 
-	trigger_create_all_on_chunk(ht, chunk);
+	if (!OidIsValid(chunk->table_id))
+		elog(ERROR, "Could not create chunk table");
 
-	/* Create all indexes on the chunk */
-	chunk_index_create_all(ht->fd.id, ht->main_table_relid, chunk->fd.id, chunk->table_id);
+	/* Create the chunk's constraints, triggers, and indexes */
+	chunk_constraints_create(chunk->constraints,
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c]: Previously, we created "dimensional" constraints separately from "inherited" constraints. Dimensional constraints were added when the chunk table was created (basically here), while inherited constraints were added after inherited triggers and indexes. Now all constraints, no matter the type, are added after table creation. If you believe it matters (I don't see why it would myself), I can move the constraints to the end. But that will instead change the order of the dimensional constraints.

File: src/chunk.c [hunk 8]
```diff
@@ -294,18 +328,74 @@ chunk_collision_resolve(Hyperspace *hs, Hypercube *cube, Point *p)
...
+		.inhRelations = list_make1(makeRangeVar(NameStr(ht->fd.schema_name), NameStr(ht->fd.table_name), 0)),
+		.tablespacename = hypertable_select_tablespace(ht, chunk),
+	};
+	Oid			uid,
+				saved_uid;
+
+	rel = heap_open(ht->main_table_relid, AccessShareLock);
+
+	if (namestrcmp(&chunk->fd.schema_name, INTERNAL_SCHEMA_NAME) == 0)
+		uid = catalog->owner_uid;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c]: I am not sure we want anyone to be able to create any type of tables in the internal schema, because users can then do a manual `CREATE TABLE` in that schema, which I don't think they should be able to do by default.

What this logic does is the following:

1. if the chunk is created in the internal schema, we create it with the permissions of the catalog/schema owner. (Still, the chunk's owner will be the owner of the hypertable.)
2. if the chunk is created in a user-specified "associated schema", then we shouldn't use the catalog owner to create the table since that typically implies super-user permissions. If we allow that, anyone can specify someone else's schema in `create_hypertable()` and start creating chunks in it without having actual permissions on the schema to do so. With this logic, the hypertable owner must have permissions to create tables in the associated schema, or else table creation will fail. If the schema doesn't yet exist, the table owner need the proper permissions on the database to create the schema.

File: src/chunk_constraint.c [hunk 9]
```diff
@@ -1,34 +1,306 @@
...
+			snprintf(NameStr(cc->fd.constraint_name),
+					 NAMEDATALEN,
+					 "constraint_%d",
+					 cc->fd.dimension_slice_id);
+			namestrcpy(&cc->fd.hypertable_constraint_name, "");
+		}
+		else if (NULL != hypertable_constraint_name)
+		{
+			CatalogSecurityContext sec_ctx;
+			char		constrname[100];
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_constraint.c]: The static analyzer of the compiler will complain in `sprintf()` that the last arg (`hypertable_constraint_name`) might not fit since the serial is an `int64`. This is to silence that warning. The name will be truncated to NAMEDATALEN by `namestrcpy()`.

I would actually prefer we used the internal constraint naming machinery (`ChooseConstraintName()`, which is similar to `makeObjectName()`), but I figured I shouldn't change naming schemes, at least not for this PR.

[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_constraint.c]: Btw, I am actually not sure why we need the `chunk_constraint` sequence at all. It seems sufficient to just have `chunk_id` and `constraint name` (i.e., it is a chunk-unique name). Then do conflict resolution by appending a unique label at the end (provided by both `makeObjectName() and `ChooseConstraintName()`). Now we don't have any conflict resolution at all.

File: src/chunk_constraint.c [hunk 10]
```diff
@@ -1,34 +1,306 @@
...
+static Oid
+chunk_constraint_create_on_table(ChunkConstraint *cc, Oid chunk_oid)
+{
+	HeapTuple	tuple;
+	Datum		values[Natts_chunk_constraint];
+	bool		nulls[Natts_chunk_constraint] = {false};
+	CatalogSecurityContext sec_ctx;
+	Relation	rel;
+
+	values[Anum_chunk_constraint_chunk_id - 1] = Int32GetDatum(cc->fd.chunk_id);
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_constraint.c]: Done.

File: src/chunk_constraint.c [hunk 11]
```diff
@@ -227,64 +475,163 @@ chunk_constraint_need_on_chunk(Form_pg_constraint conform)
...
+	heap_deform_tuple(ti->tuple, ti->desc, values, nulls);
+
+	if (nulls[Anum_chunk_constraint_hypertable_constraint_name - 1])
+		return false;
+
+	chunk_id = DatumGetInt32(values[Anum_chunk_constraint_chunk_id - 1]);
+	constrname = NameStr(*DatumGetName(values[Anum_chunk_constraint_hypertable_constraint_name - 1]));
+
+	return info->chunk_id == chunk_id &&
+		strcmp(info->hypertable_constraint_name, constrname) == 0;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_constraint.c]: That scan/delete function is not using this tuple filter, so it won't be NULL. But I guess it doesn't hurt to do an extra NULL check here.

**Comment by erimatnor** [comment]
@cevian Regarding arrays vs. lists. This is a good question. I generally prefer arrays because of the simple semantics and direct indexing. Although maybe less important, they also provide better cache locality so I make a habit of using arrays for "materialized" data. For temporary things (e.g., stuff used during computation of another final value/result) I think lists make a lot of sense. However, when stuff goes into a more persistent/cached object, I think arrays are better, can usually be more easily freed, often more space efficient (no back and forward pointers), etc.

Also, looking at tablespaces, for instance, the code now uses module arithmetic to index/select a tablespace, which obviously wouldn't work as well with lists.

I am not against a more general ADT, but it would involve a bunch of casting, etc.

File: test/expected/create_hypertable.out (hunk 3 above)
Comment (erimatnor) [reply test/expected/create_hypertable.out]: Above I test the case of having one user create chunks in another user's (associated) schema without appropriate permission. This required switching to another role. Figured it didn't matter which of the non-superusers that the rest of the tests run as. Switching back would be an extra line adding nothing to the tests just to make this line the same as before.

If you think it matters or is important I can add a `SET ROLE` above.

File: src/chunk.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/chunk.c]: Can you put a comment explaining this?

File: src/chunk.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/chunk.c]: Ah ok, that makes sense.

File: src/chunk_constraint.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/chunk_constraint.c]: the sequence assures no conflict, but I agree, let's consider using another naming scheme in another PR.

[the remark this answers is not in the document]
Comment (cevian) [reply src/chunk_constraint.c]: I don't see why changing 100->64 should give a warning, but if it does let's leave it.

File: src/process_utility.c (hunk 2 above)
Comment (cevian) [reply src/process_utility.c]: ok

File: src/process_utility.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/process_utility.c]: I would actually rather couple the catalog changes with the actual table changes to prevent the two becoming out of sync. But I'll leave this up to you.

File: test/expected/create_hypertable.out (hunk 3 above)
Comment (cevian) [reply test/expected/create_hypertable.out]: OK

File: src/chunk_constraint.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_constraint.c]: It does, as explained above. This is the warning: 
```
arning: ‘%s’ directive output may be truncated writing up to 63 bytes into a region of size between 31 and 60 [-Wformat-truncation=]
       "%d_" INT64_FORMAT "_%s"
```

File: src/chunk.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c]: done

**Comment by erimatnor** [comment]
@cevian Updated. Removed the extra commit and will submit as separate PR. Note, however, I had to add another commit to fix a broken test from the previously merged relcache leak PR. Apparently, the COPY test now had an ambiguous sort order, which made the test break on my machine. I fixed the test to have a more predictable sort order.
